api/api.proto

```diff
@@ -2684,3 +2686,74 @@ message LocateResponse {
     // <containerid>: /var/www
     map<string, string> dockerids = 2;
 }
+
+message VolumePlacementSpec {
+  // Defines a list of rules as part of the placement spec
+  repeated VolumePlacementRule rules = 1;
+}
+
+message VolumePlacementRule {
+  // Defines the number of volume replicas affected by this rule. If not provided,
+  // rule would affect all replicas
+  // +optional
+  uint64 affectedReplicas = 1;
+  // Defines the affinity rule for the volume placement
+  VolumePlacementAffinity affinity = 2;
+}
+
+// VolumePlacementAffinity allows to specify affinity and anti-affinity rules based on
+// storage pools, nodes and volumes.
+message VolumePlacementAffinity {
+  // Defines the storage pool affinity for volume placement
+  AffinityRule poolAffinity = 1;
+  // Defines the storage pool anti-affinity for volume placement
+  AffinityRule poolAntiAffinity = 2;
+
+  // Defines the node affinity for volume placement
+  AffinityRule nodeAffinity = 3;
+  // Defines the node anti-affinity for volume placement
+  AffinityRule nodeAntiAffinity = 4;
+
+  // Defines the volume affinity for volume placement
+  AffinityRule volumeAffinity = 5;
+  // Defines the volume anti-affinity for volume placement
+  AffinityRule volumeAntiAffinity = 6;
+}
+
+message AffinityRule {
+  // If false, the affinity rule is best-effort/preferred rather than enforced
+  bool bestEffort = 1;
+  // A label query over storage pools to select
+  LabelSelector labelSelector = 2;
+}
+
+
+// A label selector is a label query over a set of resources. The result of matchLabels and
+// matchExpressions are ANDed. An empty label selector matches all objects. A null
+// label selector matches no objects.
+message LabelSelector {
+  // matchLabels is a map of {key,value} pairs. A single {key,value} in the matchLabels
+  // map is equivalent to an element of matchExpressions, whose key field is "key", the
+  // operator is "In", and the values array contains only "value". The requirements are ANDed.
+  map<string, string> matchLabels = 1;
+
+  // matchExpressions is a list of label selector requirements. The requirements are ANDed.
+  repeated LabelSelectorRequirement matchExpressions = 2;
+}
+
+// A label selector requirement is a selector that contains values, a key, and an operator that
+// relates the key and values.
+message LabelSelectorRequirement {
+  // key is the label key that the selector applies to.
+  string key = 1;
+
+  // operator represents a key's relationship to a set of values.
+  // Valid operators are In, NotIn, Exists and DoesNotExist.
+  string operator = 2;
```

Comment:
This should be an enum. No string command in gRPC/protobuf :) . Add as an enum embedded in this message